layout/plan.h

```diff
@@ -0,0 +1,461 @@
+ *
+ * The directed graph is not presented as a first-class entity. Instead, a plan
+ * produces, via calls to m0_layout_plan_get(), plops as they become "ready",
+ * in the sense that all their dependencies are satisfied.
+ *
+ * Interaction between the implementation and the user takes form of the loop
+ * (which is, basically, a graph traversal):
+ *
+ *   - the user calls m0_layout_plan_get() to obtain a ready plop. The
+ *     implementation returns a plop, represented by a "sub-class" of
+ *     m0_layout_plop. Some plop fields are set up by the implementation
+ *     before the plop is handed to the user;
+ *
+ *   - the user processes the plop in the appropriate manner, e.g., executes
+ *     read or write operation (which, in turn, might involve layout plans)
+ *     and sets plop fields accordingly to the results of the execution (e.g.,
+ *     points plop to the data read from a cob);
+ *
+ *   - the user calls m0_layout_plop_done() to signal the implementation that
+ *     the plop has been executed. This might make more plops ready, so the
+ *     loop repeats.
+ *
+ * A plop is not necessarily immediately destroyed by the implementation after
+ * the user completes it. The implementation might keep plop alive for some
+ * time, for example, data read from a cob might be needed for a future parity
+ * calculation. When the implementation is done with the plop, it invokes
+ * m0_layout_plop_ops::po_fini() call-back that the user might optionally set
+ * before calling m0_layout_plop_done(). Between the call to
+ * m0_layout_plop_done() and invocation of ->po_fini() call-back the plop (and
+ * the user data buffers referenced by the plop fields) are shared between the
+ * user and the application and the user should neither modify nor free the plop
+ * or its data buffers.
+ *
+ * As an optimisation, implementation deterministically colours plops, in such a
+ * way that a user can improve locality of reference by assigning colours to
+ * processor cores. Colouring is optional and does not affect correctness.
+ *
+ * Let's consider, for example, the i/o read operation. The user calls
+ * m0_layout_plan_build() passing m0_op as argument describing the operation.
+ * After this we can enter the "loop" traversing the graph. m0_layout_plan_get()
+ * is called and returns M0_LAT_READ plop. In this plop user gets cob fid
+ * (at m0_layout_plop::pl_ent) and m0_layout_io_plop::iop_ext describing where
+ * the unit to be read is located at the cob. Now, having this information the
+ * user is ready to send the fop the ioservice.
+ *
+ * After data is read, the user puts it at m0_layout_io_plop::iop_data and
+ * calls m0_layout_plop_done(). One iteration of the graph traversing loop is
+ * done now. If the object consists only of a single unit, the next call to
+ * m0_layout_plan_get() would return M0_LAT_DONE indicating that the i/o m0_op
+ * is done and the plan is finished. But most of the times objects consist
```

Comment:
As appeared, there will be always `M0_LAT_OUT_READ` after `M0_LAT_READ` to make the data available for user.